src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs

```diff
@@ -0,0 +1,644 @@
+        {
+            if (array == null)
+                throw new ArgumentNullException("array");
+
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException("offset");
+
+            if (count < 0)
+                throw new ArgumentOutOfRangeException("count");
+
+            if (array.Length - offset < count)
+                throw new ArgumentException(SR.InvalidArgumentOffsetCount);
+        }
+
+        private void EnsureNotDisposed()
+        {
+            if (_stream == null)
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_StreamClosed);
+        }
+
+        private void EnsureDecompressionMode()
+        {
+            if (_mode != CompressionMode.Decompress)
+                throw new InvalidOperationException(SR.CannotReadFromDeflateStream);
+        }
+
+        private void EnsureCompressionMode()
+        {
+            if (_mode != CompressionMode.Compress)
+                throw new InvalidOperationException(SR.CannotWriteToDeflateStream);
+        }
+
+        public override Task<int> ReadAsync(Byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            EnsureDecompressionMode();
+
+            // We use this checking order for compat to earlier versions:
+            if (asyncOperations != 0)
+                throw new InvalidOperationException(SR.InvalidBeginCall);
+
+            ValidateParameters(array, offset, count);
+            EnsureNotDisposed();
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return TaskHelpers.FromCancellation<int>(cancellationToken);
+            }
+
+            Interlocked.Increment(ref asyncOperations);
+            Task<int> readTask = null;
+
+            try
+            {
+                // Try to read decompressed data in output buffer
+                int bytesRead = inflater.Inflate(array, offset, count);
+                if (bytesRead != 0)
+                {
+                    // If decompression output buffer is not empty, return immediately.
+                    return Task.FromResult(bytesRead);
+                }
+
+                if (inflater.Finished())
+                {
+                    // end of compression stream
+                    return Task.FromResult(0);
+                }
+
+                // If there is no data on the output buffer and we are not at 
+                // the end of the stream, we need to get more data from the base stream
+                readTask = _stream.ReadAsync(buffer, 0, buffer.Length, cancellationToken);
+                if (readTask == null)
+                {
+                    throw new InvalidOperationException(SR.NotReadableStream);
+                }
+
+                var tcs = new TaskCompletionSource<int>();
+
+                // ContinueWith will never throw here otherwise we'll be in inconsistent state
+                readTask.ContinueWith(
```

Comment:
It's not clear to me why ContinueWith is being used in this fashion rather than using async/await.  We should open a work item to revisit this async implementation.